src/option_learning.py

```diff
@@ -0,0 +1,190 @@
+        with each other).
+        """
+        raise NotImplementedError("Override me!")
+
+    @abc.abstractmethod
+    def update_segment_from_option_spec(
+            self, segment: Segment, option_spec: OptionSpec) -> None:
+        """Figure out which option was executed within the given segment.
+        Modify the segment in-place to include this option, via
+        segment.set_option().
+
+        At this point, we know which ParameterizedOption was used, and we know
+        the option_vars. This information is included in the given option_spec.
+        But we don't know what parameters were used in the option,
+        which this method should figure out.
+        """
+        raise NotImplementedError("Override me!")
+
+
+class _KnownOptionsOptionLearner(_OptionLearnerBase):
+    """The "option learner" that's used when we're in the code path where
+    CFG.do_option_learning is True.
+    """
+    def learn_option_specs(
+            self, strips_ops: List[STRIPSOperator],
+            partitions: List[Partition]) -> List[OptionSpec]:
+        # Since we're not actually doing option learning, the data already
+        # contains the options. So, we just extract option specs from the data.
+        option_specs = []
+        for partition in partitions:
+            param_option = None
+            option_vars = None
+            for i, (segment, sub) in enumerate(partition):
+                option = segment.actions[0].get_option()
+                if i == 0:
+                    param_option = option.parent
+                    option_vars = [sub[o] for o in option.objects]
+                else:
+                    assert param_option == option.parent
+                    assert option_vars == [sub[o] for o in option.objects]
+                # Make sure the option is consistent within a trajectory.
+                for a in segment.actions:
+                    option_a = a.get_option()
+                    assert param_option == option_a.parent
+                    assert option_vars == [sub[o] for o in option_a.objects]
+            assert param_option is not None and option_vars is not None, \
+                "No data for this partition?"
+            option_specs.append((param_option, option_vars))
+        return option_specs
+
+    def update_segment_from_option_spec(
+            self, segment: Segment, option_spec: OptionSpec) -> None:
+        # If we're not doing option learning, the segments will already have
+        # the options, so there is nothing to do here.
+        pass
+
+
+class _OracleOptionLearner(_OptionLearnerBase):
+    """The option learner that just cheats by looking up ground truth options
+    from the environment. Useful for testing.
+    """
+    def learn_option_specs(
+            self, strips_ops: List[STRIPSOperator],
+            partitions: List[Partition]) -> List[OptionSpec]:
+        ### THIS HAS ONLY BEEN TESTED FOR THE COVER AND BLOCKS ENVIRONMENTS!!!
```

Comment:
we should add and test with CoversMultistepOptions soon, since it's the only env we have right now with multistep options. do you want to do that in this PR, or wait until the next? I'm okay with waiting, but it may reveal bugs

Reply:
as discussed on slack, this is hard. requires reverse engineering the option policies which are complicated